### ISSUE TITLE
Support for OpenAPI schemas defined in components

### ISSUE THREAD

**Issue by avicennax** [bentoml/BentoML#2142]
**Is your feature request related to a problem? Please describe.**
Currently `api` endpoints can support locally defined schemas for inputs with [recently](https://github.com/bentoml/BentoML/issues/1480) added support schema for JSON outputs.

However AFIACT there no support for reusable schemas definitions via [components](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#componentsObject). Furthermore, this makes it much harder to generate specs for nested data structures such as `dataclasses` and Pydantic `BaseModel` objects since schema generation will use refs to non-local definitions (either components or definitions; latter of which cannot be used inside [MediaTypeObject](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#mediaTypeObject) where `request_schema` dumps).

**Describe the solution you'd like**
Provide something like a `BentoService.Swagger` access object that allows users to modify the generated OpenAPI spec in sane ways. Whatever makes sense to you guys tbh - just passing in the component JSON blob to an appropriate method is sufficient.

**Describe alternatives you've considered**
- Using FastAPI to hand roll the service.
- Manually update `request_schema` dumps so all refs are replaced with minimal, local, `properties` only definitions.

**Additional context**
Some of our model endpoints will receive deeply nested data inputs.

**Comment by parano** [comment]
HI @avicennax - when we were adding OpenAPI & PyDantic support recently, @aarnphm actually explored the idea of using OpenAPI components, it is completely feasible but we were not sure what the main benefits are. Could you share a bit more about your use case? What's the main benefit of using OpenAPI schema components?

Note that since we are adding support for PyDantic models for Json input & output, it is possible to create a common PyDantic class for shared attributes, among different service inputs.

**Comment by avicennax** [comment]
Sure, I'll give a specific (albeit toy) example here. Imagine our service input classes are defined via these Pydantic models:

```python
from pydantic import BaseModel


class IrisFeatures(BaseModel):
    sepal_len: float
    sepal_width: float
    petal_len: float
    petal_width: float


class IrisClassificationRequest(BaseModel):
    request_id: str
    iris_features: IrisFeatures
```

We can create an adapter class of some variant like `PydanticInput(BaseInputAdapter)` which takes our model as an input (sets it to `self.pydantic_model`) and implements `extract_user_func_args` to marshal the payload - so far all good. However, when we go implement `request_schema`, simply returning something like `{"application/json": {"schema": self.pydantic_model.schema()}}` will fail as as the schema dump assumes there is a component definition schema named `IrisFeatures` and AFAICT there isn't a way to easily fix this without manually modifying the schema dump.

Basically nested Pydantic models can't generate schema dumps that work out the box with BentoML's OpenAPI spec and likewise with Marshmellow + dataclasses (AFAICT...). While I can flatten the model, I think most observers will agree even in the toy example the nested models provide a cleaner abstraction.

So the main argument is that it will enable users to easily create custom adapters using validated, nested data models with functioning OpenAPI schemas; the latter being critical for generating out-the-box functioning code-gen clients in statically typed languages.

Let me know if this makes sense and/or if I've missed something on your end.

**Comment by avicennax** [comment]
Also, I'm happy to tank the implementation of this feature with direction from you or @aarnphm; since I assume it would be a relatively low priority if you guys decide it's worth pursuing.

**Comment by parano** [comment]
@avicennax would be great to get your help on this one - the `main` branch is for the development of 1.0, it would be great if you could help take a look at the current API IO descriptor and related OpenAPI implementation, and would be great to discuss more with you about your proposed solution.

**Comment by aarnphm** [comment]
it seems like to me that BentoML's OpenAPI schema definitions should be able to process nested pydantic schema definition. It would be great if you can take a look into this. Feel free to dm me or Chaoyu for your proposed solution.

**Comment by ahallermed** [comment]
That would be really great. At the moment, I used a workaround to overcome this issue. Maybe, this helps other folks who stumble upon the same problem but still want to have pydantic schemas used in the api:

```python
    def create_post_request_schema(cls, model: Type[pydantic.BaseModel], endpoint_name: str) -> Dict[str, Any]:
        """
        Args:
            endpoint_name: the name of the endpoint
                example: if using 'http://localhost:5000/my_prediction', then the name is: "my_prediction"

        Explanations:
        - openapi defintions can be tested by copying the docs.json into https://editor.swagger.io/
            and experiment with it
        - '/' and '~' need to be esacped to ~1 and ~0 respectively;
            Ref: https://swagger.io/docs/specification/using-ref/
        """
        ref_template = (f"#/paths/~1{endpoint_name}/post/requestBody/"
                        f"content/application~1json/schema/definitions/{{model}}")
        return model.schema(ref_template=ref_template)
```

Only tested on 0.13 but I also had a look at the current main for 1.0 and there is should work either.
Unfortunately, you also need to create such a function for 200 responses, but they can be created analogously.

**Comment by parano** [comment]
Confirmed that a nested Pydantic model doesn't work with BentoML 1.0 today, here's an example modified from the quickstart guide:

```
import numpy as np
import pandas as pd
import bentoml
from bentoml.io import NumpyNdarray, JSON
from pydantic import BaseModel

iris_clf_runner = bentoml.sklearn.get("iris_clf:latest").to_runner()

svc = bentoml.Service("iris_classifier", runners=[iris_clf_runner])

class IrisFeatures(BaseModel):
    sepal_len: float
    sepal_width: float
    petal_len: float
    petal_width: float


class IrisClassificationRequest(BaseModel):
    request_id: str
    iris_features: IrisFeatures

input_spec = JSON(pydantic_model=IrisClassificationRequest)

@svc.api(input=input_spec, output=NumpyNdarray())
def classify(input_data: IrisClassificationRequest) -> np.ndarray:
    print("request_id:", input_data.request_id)
    input_df = pd.DataFrame(input_data.iris_features.dict())
    return iris_clf_runner.predict.run(input_df)
```

When starting up the server and visit swagger UI, it shows the following error:

<img width="932" alt="Screen Shot 2022-07-22 at 1 57 36 PM" src="https://user-images.githubusercontent.com/489344/180567511-19dc50c1-58fa-4803-a297-a1f92e788e69.png">


Below is the generated openAPI doc:

```
{
  "openapi": "3.0.0",
  "info": {
    "title": "iris_classifier",
    "description": "A Prediction Service built with BentoML",
    "contact": {
      "email": "contact@bentoml.ai"
    },
    "version": "0.0.0"
  },
  "tags": [
    {
      "name": "infra",
      "description": "Infrastructure endpoints"
    },
    {
      "name": "app",
      "description": "Inference endpoints"
    }
  ],
  "paths": {
    "/healthz": {
      "get": {
        "tags": [
          "infra"
        ],
        "description": "Health check endpoint. Expecting an empty response with status code <code>200</code> when the service is in health state. The <code>/healthz</code> endpoint is <b>deprecated</b> (since Kubernetes v1.16)",
        "responses": {
          "200": {
            "description": "success"
          }
        }
      }
    },
    "/livez": {
      "get": {
        "tags": [
          "infra"
        ],
        "description": "Health check endpoint for Kubernetes. Healthy endpoint responses with a <code>200</code> OK status.",
        "responses": {
          "200": {
            "description": "success"
          }
        }
      }
    },
    "/readyz": {
      "get": {
        "tags": [
          "infra"
        ],
        "description": "A <code>200</code> OK status from <code>/readyz</code> endpoint indicated the service is ready to accept traffic. From that point and onward, Kubernetes will use <code>/livez</code> endpoint to perform periodic health checks.",
        "responses": {
          "200": {
            "description": "success"
          }
        }
      }
    },
    "/metrics": {
      "get": {
        "tags": [
          "infra"
        ],
        "description": "Prometheus metrics endpoint",
        "responses": {
          "200": {
            "description": "success"
          }
        }
      }
    },
    "/classify": {
      "post": {
        "tags": [•],
        "summary": "InferenceAPI(JSON(pydantic_model=<class 'service.IrisClassificationRequest'>) â†’ NumpyNdarray())",
        "description": "",
        "operationId": "iris_classifier__classify",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "title": "IrisClassificationRequest",
                "type": "object",
                "properties": {
                  "request_id": {
                    "title": "Request Id",
                    "type": "string"
                  },
                  "iris_features": {
                    "$ref": "#/definitions/IrisFeatures"
                  }
                },
                "required": [
                  "request_id",
                  "iris_features"
                ],
                "definitions": {
                  "IrisFeatures": {
                    "title": "IrisFeatures",
                    "type": "object",
                    "properties": {
                      "sepal_len": {
                        "title": "Sepal Len",
                        "type": "number"
                      },
                      "sepal_width": {
                        "title": "Sepal Width",
                        "type": "number"
                      },
                      "petal_len": {
                        "title": "Petal Len",
                        "type": "number"
                      },
                      "petal_width": {
                        "title": "Petal Width",
                        "type": "number"
                      }
                    },
                    "required": [
                      "sepal_len",
                      "sepal_width",
                      "petal_len",
                      "petal_width"
                    ]
                  }
                }
              }
            }
          }
        },
        "responses": {
          "200": {
            "description": "success",
            "content": {
              "application/json": {
                "schema": {
                  "type": "array",
                  "items": {}
                }
              }
            }
          },
          "400": {
            "description": "Bad Request"
          },
          "404": {
            "description": "Not Found"
          },
          "500": {
            "description": "Internal Server Error"
          }
        }
      }
    }
  }
}
```

We should add the definitions part from Pydantic schema output to under `/components/schema/` of the generated openAPI doc. cc @aarnphm

**Comment by aarnphm** [comment]
Development progress will be tracked at #2845 and hopefully will be included in the upcoming release.
